caddy/ws2outline_handler.go

```diff
@@ -0,0 +1,187 @@
+
+import (
+	"errors"
+	"fmt"
+	"log/slog"
+	"net"
+	"net/http"
+
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+	"github.com/caddyserver/caddy/v2"
+	"github.com/caddyserver/caddy/v2/modules/caddyhttp"
+	"github.com/mholt/caddy-l4/layer4"
+	"go.uber.org/zap"
+	"golang.org/x/net/websocket"
+)
+
+const (
+	wsModuleName = "http.handlers.ws2outline"
+)
+
+func init() {
+	caddy.RegisterModule(ModuleRegistration{
+		ID:  wsModuleName,
+		New: func() caddy.Module { return new(WebSocketHandler) },
+	})
+}
+
+type ConnectionType string
+
+const (
+	connectionTypeStream ConnectionType = "stream"
+	connectionTypePacket ConnectionType = "packet"
+)
+
+// WebSocketHandler implements a middleware Caddy web handler that proxies
+// WebSockets Outline connections.
+type WebSocketHandler struct {
+	// The type of connection.
+	Type              ConnectionType `json:"type,omitempty"`
+	ConnectionHandler string         `json:"connection_handler,omitempty"`
+	compiledHandler   layer4.NextHandler
+
+	logger  *slog.Logger
+	zlogger *zap.Logger
+}
+
+var (
+	_ caddy.Provisioner           = (*WebSocketHandler)(nil)
+	_ caddy.Validator             = (*WebSocketHandler)(nil)
+	_ caddyhttp.MiddlewareHandler = (*WebSocketHandler)(nil)
+)
+
+func (*WebSocketHandler) CaddyModule() caddy.ModuleInfo {
+	return caddy.ModuleInfo{ID: wsModuleName}
+}
+
+// Provision implements caddy.Provisioner.
+func (h *WebSocketHandler) Provision(ctx caddy.Context) error {
+	h.logger = ctx.Slogger()
+	h.zlogger = ctx.Logger()
+	if h.Type == "" {
+		// The default connection type if not provided is a stream.
+		h.Type = connectionTypeStream
+	}
+
+	mod, err := ctx.AppIfConfigured(outlineModuleName)
+	if err != nil {
+		return fmt.Errorf("outline app configure error: %w", err)
+	}
+	app, ok := mod.(*OutlineApp)
+	if !ok {
+		return fmt.Errorf("module `%s` is of type `%T`, expected `OutlineApp`", outlineModuleName, app)
+	}
+	for _, compiledHandler := range app.Handlers {
+		if compiledHandler.Name == h.ConnectionHandler {
+			h.compiledHandler = compiledHandler
+			break
+		}
+	}
+	if h.compiledHandler == nil {
+		return fmt.Errorf("no connection handler configured for `%s`", h.ConnectionHandler)
+	}
+
+	return nil
+}
+
+// Validate implements caddy.Validator.
+func (h *WebSocketHandler) Validate() error {
+	if h.Type != "" && h.Type != connectionTypeStream && h.Type != connectionTypePacket {
+		return fmt.Errorf("unsupported `type`: %v", h.Type)
+	}
+	if h.ConnectionHandler == "" {
+		return errors.New("must specify `connection_handler`")
+	}
+
+	return nil
+}
+
+// ServeHTTP implements caddyhttp.MiddlewareHandler.
+func (h WebSocketHandler) ServeHTTP(w http.ResponseWriter, r *http.Request, _ caddyhttp.Handler) error {
+	h.logger.Debug("handling connection",
+		slog.String("path", r.URL.Path),
+		slog.Any("remote_addr", r.RemoteAddr))
+
+	switch h.Type {
+	case connectionTypeStream:
+		handler := func(wsConn *websocket.Conn) {
+			raddr, err := net.ResolveTCPAddr("tcp", r.RemoteAddr)
```

Comment:
Why do you need to resolve the address? Same for UDP.

Also, where do you extract the "forwarded for" header and pass to the handler? We need that logic.